source/Octopus.Cli/Commands/Releases/ReleasePlanBuilder.cs

```diff
@@ -24,54 +29,122 @@ public ReleasePlanBuilder(ILogger log, IPackageVersionResolver versionResolver,
             this.commandOutputProvider = commandOutputProvider;
         }
 
-        public async Task<ReleasePlan> Build(IOctopusAsyncRepository repository, ProjectResource project, ChannelResource channel, string versionPreReleaseTag)
+        public async Task<ReleasePlan> Build(IOctopusAsyncRepository repository, ProjectResource project,
+            ChannelResource channel, string versionPreReleaseTag, string gitReference)
+        {
+            return string.IsNullOrWhiteSpace(gitReference)
+                ? await BuildReleaseFromDatabase(repository, project, channel, versionPreReleaseTag)
+                : await BuildReleaseFromVersionControl(repository, project, channel, versionPreReleaseTag, gitReference);
+        }
+
+        async Task<ReleasePlan> BuildReleaseFromVersionControl(IOctopusAsyncRepository repository, ProjectResource project,
+            ChannelResource channel, string versionPreReleaseTag, string gitReference)
+        {
+            if (repository == null) throw new ArgumentNullException(nameof(repository));
+            if (project == null) throw new ArgumentNullException(nameof(project));
+            if (!project.IsVersionControlled)
+            {
+                throw new CommandException(GitReferenceSuppliedForDatabaseProjectErrorMessage(gitReference));
+            }
+            commandOutputProvider.Debug($"Finding deployment process at git reference {gitReference}...");
+            var deploymentProcess = await repository.DeploymentProcesses.Beta().Get(project, gitReference);
```

Comment:
Just a thought: 
feels strange to me that we use the ref to get the deployment process for the release plan, and yet we create the release by resolving a commit at release creation time. Which would mean the deployment process could be different. between the release plan and release creation

But this is a symptom of how create release using the api, as we ask for ref instead of commit. I guess the release plan is mainly here for validation. Perhaps we could have create release dry run api, which would give us the deployment process to use in the plan.

Reply:
curious to know how we handle it now with database projects. if someone changes the process in between release plan and release creation

Reply:
Yeah, we have the same potential for problems with database projects. It seems that version controlled projects might be more prone to the process changing between "get template" and "create release"? Or maybe commits are just more obvious than database transaction?

Reply:
Yeah I think commits definitely make it more obvious, the whole idea around release plan or release template, is a concern for our CLI or UI right?. perhaps we should resolve the commit from the ref first and then internally we use the commit as the ref. Where as customers can passing either commit or other refs 🤷‍♂️